ios/ReactNativePageView.m

```diff
@@ -218,307 +40,299 @@ - (instancetype)initWithEventDispatcher:(RCTEventDispatcher *)eventDispatcher {
     }
 }
 
-- (void)pageViewController:(UIPageViewController *)pageViewController
-        didFinishAnimating:(BOOL)finished
-   previousViewControllers: (nonnull NSArray<UIViewController *> *)previousViewControllers
-       transitionCompleted:(BOOL)completed {
-    if (completed) {
-        UIViewController* currentVC = pageViewController.viewControllers[0];
-        _currentIndex = [_childrenViewControllers indexOfObject:currentVC];
-        [_eventDispatcher sendEvent:[[RCTOnPageSelected alloc] initWithReactTag:self.reactTag position:[NSNumber numberWithInteger:_currentIndex] coalescingKey:_coalescingKey++]];
-
-        [_eventDispatcher sendEvent:[[RCTOnPageScrollEvent alloc] initWithReactTag:self.reactTag position:[NSNumber numberWithInteger:_currentIndex] offset:[NSNumber numberWithFloat:0]]];
-        _reactPageIndicatorView.currentPage = _currentIndex;
+- (void)goTo:(NSInteger)index animated:(BOOL)animated {
+    NSInteger numberOfPages = self.reactSubviews.count;
+    
+    if (numberOfPages == 0 || index < 0) {
+        return;
     }
-}
+        
+    UIPageViewControllerNavigationDirection direction = (index > self.currentIndex) ? UIPageViewControllerNavigationDirectionForward : UIPageViewControllerNavigationDirectionReverse;
+    
+    NSInteger indexToDisplay = index < numberOfPages ? index : numberOfPages - 1;
+    
+    UIView *viewToDisplay = self.reactSubviews[indexToDisplay];
+    UIViewController *controllerToDisplay = [self findAndCacheControllerForView:viewToDisplay];
 
-#pragma mark - Datasource After
+    self.reactPageIndicatorView.numberOfPages = numberOfPages;
+    self.reactPageIndicatorView.currentPage = indexToDisplay;
+        
+    [self setReactViewControllers:indexToDisplay
+                             with:controllerToDisplay
+                        direction:direction
+                         animated:animated];
+    
+}
 
-- (UIViewController *)pageViewController:
-(UIPageViewController *)pageViewController
-       viewControllerAfterViewController:(UIViewController *)viewController {
-    NSUInteger index = [_childrenViewControllers indexOfObject:viewController];
+- (UIViewController *)findAndCacheControllerForView:(UIView *)viewToDisplay {
+    if (!viewToDisplay) { return nil; }
     
-    if (index == NSNotFound) {
-        return nil;
+    UIViewController *controllerToDisplay = [self findCachedControllerForView:viewToDisplay];
+    UIViewController *current = [self currentlyDisplayed];
+
+    if (!controllerToDisplay && current.view.reactTag == viewToDisplay.reactTag) {
+        controllerToDisplay = current;
     }
-    
-    index++;
-    
-    if (index == [_childrenViewControllers count]) {
-        return nil;
+    if (!controllerToDisplay) {
+         controllerToDisplay = [[UIViewController alloc] initWithView:viewToDisplay];
     }
-    return [_childrenViewControllers objectAtIndex:index];
+    [self.cachedControllers addObject:controllerToDisplay];
```

Comment:
should we always add controller to `cachedControllers`
IMO it should be only, when `controllerToDisplay` is null

Reply:
I am doing that just to be sure to have all controllers cached. They are not retained so no risk